### PR TITLE
960 Client Credentials Grant Request builders

### LINKED ISSUE #960: Missing Client Credentials request builder
A password grant request is written like this:
```
OAuthPasswordGrantRequestAuthentication createRequest = OAuthRequests.OAUTH_PASSWORD_GRANT_REQUEST.builder().setLogin(email).setPassword("Change&45+me1!").build();
OAuthGrantRequestAuthenticationResult result = Authenticators.OAUTH_PASSWORD_GRANT_REQUEST_AUTHENTICATOR.forApplication(app).authenticate(createRequest)
```
But a Client Credentials request is written like this:
```
OAuthClientCredentialsGrantRequestAuthentication clientCredentialsGrantRequestAuthentication =
new  DefaultOAuthClientCredentialsGrantRequestAuthentication(apiKeyId, apiKeySecret);
authenticationResult = Authenticators.OAUTH_CLIENT_CREDENTIALS_GRANT_REQUEST_AUTHENTICATOR
                    .forApplication(app)
                    .authenticate(clientCredentialsGrantRequestAuthentication);
```

The latter is missing a builder to create the request. Users are forced to instantiate the `                    DefaultOAuthClientCredentialsGrantRequestAuthentication` themselves. Users will need to depend on `com.stormpath.sdk:stormpath-sdk-impl` because of this. This goes against our api/impl separation strategy. 

A builder for client credentials is required in `OAuthRequests`

### PR DESCRIPTION
**Pull request by mzumbadoate**
fixes #960

### CODE REVIEW

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthClientCredentialsGrantRequestAuthenticationFactory.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/*
+* Copyright 2015 Stormpath, Inc.
```
Comment (mrioan) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthClientCredentialsGrantRequestAuthenticationFactory.java]: 2016

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthClientCredentialsGrantRequestAuthenticationBuilder.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
+/*
+* Copyright 2015 Stormpath, Inc.
```
Comment (mrioan) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthClientCredentialsGrantRequestAuthenticationBuilder.java]: 2016

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthRequests.java [hunk 3]
```diff
@@ -69,4 +69,14 @@ private OAuthRequests() {
      * @since 1.0.RC8.2
      */
     public static final IdSiteAuthenticationRequestFactory IDSITE_AUTHENTICATION_REQUEST = (IdSiteAuthenticationRequestFactory) Classes.newInstance("com.stormpath.sdk.impl.oauth.DefaultIdSiteAuthenticationRequestFactory");
+
+    /**
+     * Returns a new {@link OAuthClientCredentialsRequestAuthenticatorFactory} instance, used to construct Create Client Credentials Authentication requests.
+     *
+     * @return a new {@link OAuthClientCredentialsRequestAuthenticatorFactory} instance, used to construct Create Client Credentials Authentication requests.
+     *
+     * @since 1.1.0
```
Comment (mrioan) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthRequests.java]: 1.0.4

File: impl/src/main/java/com/stormpath/sdk/impl/oauth/DefaultOAuthClientCredentialsGrantRequestAuthenticationBuilder.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
+/*
+* Copyright 2015 Stormpath, Inc.
```
Comment (mrioan) [inline impl/src/main/java/com/stormpath/sdk/impl/oauth/DefaultOAuthClientCredentialsGrantRequestAuthenticationBuilder.java]: 2016

File: impl/src/main/java/com/stormpath/sdk/impl/oauth/DefaultOAuthClientCredentialsGrantRequestAuthenticationFactory.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+/*
+* Copyright 2015 Stormpath, Inc.
```
Comment (mrioan) [inline impl/src/main/java/com/stormpath/sdk/impl/oauth/DefaultOAuthClientCredentialsGrantRequestAuthenticationFactory.java]: 2016

**Comment by mrioan** [comment]
In `ApplicationIT.groovy` we have many tests that validate that the each Oauth authentication works ok. Can you please add tests for client credentials there? Thanks. Back to you.

**Comment by mzumbadoate** [comment]
@mrioan I fixed the javadocs and also added a test to ApplicationIT
ready for a new round of reviews

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/ApplicationIT.groovy [hunk 6]
```diff
@@ -1654,6 +1655,29 @@ class ApplicationIT extends ClientIT {
...
+        RequestCountingClient client = buildCountingClient();
+
+        app = client.getResource(app.href, Application)
+
+        def appApiKey = app.getApiKey(apiKey.id, ApiKeys.options().withAccount().withTenant())
+
+        OAuthClientCredentialsGrantRequestAuthentication ccrequest = OAuthRequests.OAUTH_CLIENT_CREDENTIALS_GRANT_REQUEST.builder().setApiKeyId(appApiKey.id).setApiKeySecret(appApiKey.secret).build();
+        def result = Authenticators.OAUTH_CLIENT_CREDENTIALS_GRANT_REQUEST_AUTHENTICATOR.forApplication(app).authenticate(ccrequest)
+
+        assertNotNull result
```
Comment (mrioan) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/ApplicationIT.groovy]: @mzumbadoate  Asserting the result is not null is not enough. You should also assert the expected values of the result, like the account for example.

**Comment by mrioan** [comment]
@mzumbadoate I just update the IT. It fails because the refresh token is null. That seems like  bug to me (unless client credentials are not supposed to have a refresh token which I strongly doubt). Back to you.

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/ApplicationIT.groovy [hunk 7]
```diff
@@ -1666,16 +1667,20 @@ class ApplicationIT extends ClientIT {
...
+        OAuthClientCredentialsGrantRequestAuthentication request = OAuthRequests.OAUTH_CLIENT_CREDENTIALS_GRANT_REQUEST.builder().setApiKeyId(apiKey.id).setApiKeySecret(apiKey.secret).build();
+        def result = Authenticators.OAUTH_CLIENT_CREDENTIALS_GRANT_REQUEST_AUTHENTICATOR.forApplication(app).authenticate(request)
 
-        app = client.getResource(app.href, Application)
+        assertNotNull result.getAccessTokenHref()
+        assertEquals result.getAccessToken().getHref(), result.getAccessTokenHref()
+        assertEquals(result.getAccessToken().getAccount().getHref(), account.getHref())
+        assertEquals(result.getAccessToken().getApplication().getHref(), app.getHref())
+        assertTrue Strings.hasText(result.getAccessTokenString())
+        assertEquals(result.getRefreshToken().getAccount().getHref(), account.getHref())
```
Comment (mrioan) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/ApplicationIT.groovy]: @mzumbadoate Error here as refresh token is null

**Comment by mzumbadoate** [comment]
@mrioan 
In removing refresh token validation from the IT because of this
https://tools.ietf.org/html/rfc6749#section-4.4.3
"A refresh token SHOULD NOT be included."

back to you
since I'm off on PTO I won't be able to merge this

**Comment by mrioan** [comment]
@jarias I have updated this PR myself. @mzumbadoate was the original implementer but he is out on vacation. Could you please review it? Thanks!

**Comment by jarias** [comment]
LGTM @mrioan

**Comment by dogeared** [comment]
having trouble with github's review feature. keeps giving me a 500 error. Why are we hardcoding `4.1.1` for clover and deleting `clover.version`?

**Pull request merged**